File: src/com/gmail/marcosav2010/peer/PeerThreadFactory.java
package com.gmail.marcosav2010.peer;

import java.util.concurrent.ThreadFactory;
import java.util.concurrent.atomic.AtomicInteger;

class PeerThreadFactory implements ThreadFactory {

    private final ThreadGroup group;
    private final AtomicInteger threadNumber = new AtomicInteger(1);
    private final String namePrefix;

    PeerThreadFactory(String peerName) {
        this(peerName, null);
    }

    PeerThreadFactory(String peerName, ThreadGroup parentThreadGroup) {
        String threadGroupName = "peerThreadGroup-" + peerName;
        if (parentThreadGroup != null)
            group = new ThreadGroup(parentThreadGroup, threadGroupName);
        else
            group = new ThreadGroup(threadGroupName);

        namePrefix = "peerPool-" + peerName + "-thread-";
    }

    @Override
    public Thread newThread(Runnable r) {
        Thread t = new Thread(group, r, namePrefix + threadNumber.getAndIncrement(), 0);
        if (t.isDaemon())
            t.setDaemon(false);
        if (t.getPriority() != Thread.NORM_PRIORITY)
            t.setPriority(Thread.NORM_PRIORITY);

        return t;
    }
}

File: core/src/main/java/com/gmail/marcosav2010/peer/PeerThreadFactory.kt
package com.gmail.marcosav2010.peer

import java.util.concurrent.ThreadFactory
import java.util.concurrent.atomic.AtomicInteger

internal class PeerThreadFactory @JvmOverloads constructor(peerName: String, parentThreadGroup: ThreadGroup? = null) :
        ThreadFactory {

    private var group: ThreadGroup? = null
    private val threadNumber = AtomicInteger(1)
    private val namePrefix: String

    init {
        val threadGroupName = "peerThreadGroup-$peerName"
        group = parentThreadGroup?.let { ThreadGroup(it, threadGroupName) } ?: ThreadGroup(threadGroupName)
        namePrefix = "peerPool-$peerName-thread-"
    }

    override fun newThread(r: Runnable): Thread {
        val t = Thread(group, r, namePrefix + threadNumber.getAndIncrement(), 0)
        if (t.isDaemon) t.isDaemon = false
        if (t.priority != Thread.NORM_PRIORITY) t.priority = Thread.NORM_PRIORITY
        return t
    }
}